### DISCUSSION TITLE
Accessing inner node by reference

### DISCUSSION THREAD

**Discussion by yoav-spring** [nlohmann/json#3940] [Q&A]
I have a JSON document d where an inner node contains counters I need to add / increment, e.g.:

```
json d = {
    "foo": "bar",
    "MyComponent": {
        "Counters": {
            "1": 12,
            "10": 8,
            "11": 6,
            "12": 7,
            "2": 11,
            "3": 10,
            "4": 9,
            "5": 6,
            "6": 11,
            "7": 10,
            "8": 9,
            "9": 7
        }
    }
}
```

I would like to initialize and maintain a reference `counters` to d["MyComponent"]["Counters"] that I can read but also write to, affecting d. I tried accessing the counters map using at(), but writing to it didn't change the original document.

Is there a way to achieve that?

**Comment by nlohmann** [comment]
at should work as it returns a reference. Can you please share your code?

**Reply by yoav-spring** [reply]
Thanks for the prompt reply.

I start with an empty Counters element.
```
    string key = to_string(42);
    json d = TelemetryLogger::Get(); // a singleton supplying json&
    // auto counters = d["MyComponent"]["Counters"]; // this works, next line doesn't
    auto counters = d.at("/MyComponent/Counters"_json_pointer);
    int counter = !counters.contains(key) ? 0 : counters[key].get<int>(); // check if counter exists, else initialize to zero
    // d["MyComponent"]["Counters"][key] = counter + 1; // this works, next line doesn't
    counters[key] = counter + 1;
```
At the end of the run _d_ stays unchanged, while _counters_ is modified.
I also get a deprecation warning when using at() with a pointer string. I saw someone mentioning this in one of the discussions:
```
In file included from ../src/point-to-point/model/telemetry-logger.h:1:0,
                 from ../src/point-to-point/model/reordering-buffer.cc:1:
../nlohmann/json.hpp:13800:5: note: declared here
     operator string_t() const
     ^
In file included from ../src/point-to-point/model/telemetry-logger.h:1:0,
                 from ../src/point-to-point/model/reordering-buffer.cc:1:
../nlohmann/json.hpp:3913:33: warning: ‘nlohmann::json_abi_v3_11_2::json_pointer<RefStringType>::operator nlohmann::json_abi_v3_11_2::json_pointer<RefStringType>::string_t() const [with RefStringType = std::__cxx11::basic_string<char>; nlohmann::json_abi_v3_11_2::json_pointer<RefStringType>::string_t = std::__cxx11::basic_string<char>]’ is deprecated: Since 3.11.0; use to_string() [-Wdeprecated-declarations]
 decltype(std::declval<Compare>()(std::declval<B>(), std::declval<A>()))
                                 ^
```

**Reply by yoav-spring** [reply]
I think I realized what I've been doing wrong - once I replaced:
`    auto counters = d.at("/MyComponent/Counters"_json_pointer);`
with:
`    json& counters = d.at("/MyComponent/Counters"_json_pointer);`
The original document changed as well!

**Reply by nlohmann** [reply]
Oh yes. Without the & you just made a copy. Hope this fixes it for you.

**Reply by yoav-spring** [reply]
Do know how to avoid the deprecation warning (apart from suppressing it via compiler flags / pragma)? I'm using gcc / g++ 5.5 and json 3.11.2.

**Reply by nlohmann** [reply]
Which line in your code triggers the message? See https://json.nlohmann.me/integration/migration_guide/ for information on deprecated functions.
